posts/coding-intentionally-in-bash-grains.md

```diff
@@ -0,0 +1,194 @@
+
+```bash
+total=0
+current_grains=1
+for square in {1..64}; do
+  total=$( bc <<< "$total + $current_grains" )
+  current_grains=$( bc <<< "$current_grains * 2" )
+done
+echo "$total"
+```
+
+That's a lot more readable and understandable.  The code clearly shows that the number of squares on the chess board is a driving factor, as well as the doubling effect for each square.  I think this is better than the initial solution.
+
+However.
+
+It's slow.  Looping, adding, and repeatedly calling out to an external command?  They all stack up to a slow-ish run-time.  Now, is that a huge deal?  No.  If you're scripting this in Bash, you've probably already decided you don't have speed constraints.  But could it be better?  Yes.
+
+### Third option: direct calculation
+
+So, how do we add all these up without iterating?
+
+Let's consider a *smaller* version of the same problem: a chessboard with 5 squares[^3].
+
+The five squares would have the following number of grains:
+
+```txt
+---------------------
+| 1 | 2 | 4 | 8 |16 |
+---------------------
+```
+
+And the total here would be: 1 + 2 + 4 + 8 + 16 = 31.  Hm.  31 doesn't scream anything obvious at me yet.  Let's go a little bigger.
+
+OK, well what about a 6-square chessboard?  This time, I'll show the running total below each square to help us add it up.
+
+```txt
+-------------------------
+| 1 | 2 | 4 | 8 |16 |32 |
+|   | 3 | 7 |15 |31 |63 |
+-------------------------
+```
+
+And the sum: 1 + 2 + 4 + 8 + 16 + 32 = 63.  Hmm... I'm actually starting to see a glimmer of a pattern, but we'll do one more just to be sure.
+
+7 squares:
+
+```txt
+-----------------------------
+| 1 | 2 | 4 | 8 |16 |32 |64 |
+|   | 3 | 7 |15 |31 |63 |127|
+-----------------------------
+```
+
+1 + 2 + 4 + 8 + 16 + 32 +64 = 127.  Do you see it?  Anything ringing any alarms with the values 31, 63, 127?
+
+They're *almost* powers of 2.  In fact, they're *one less* than the *next* power of two.  
+
+One more example, to drive it home.  Imagine a 12-square chess board. That's twelve doublings, or 2 multiplied by itself 12 times (which, in the math biz, is 2^12): 4096. Double that again and you get 8192 (2^13). So... if we got the pattern right, the running total would be *one short* of 8192, also known as 8191. And if we tally it up, that's, exactly what we get: 1 + 2 + 4 + 8 + 16 + 32 + 64 + 128 + 256 + 512 + 1024 + 2048 + 4096 = 8191.
+
+> To put it another way, to find the total for all `n` squares, all you need is the value on square `n + 1` minus one.
+
+Soooooo, if we want to total up all of the squares up through square 64, all we need to do is calculate the number of grains on *square 65* and subtract 1.
+
+Blammo.
+
+In Bash, to calculate the value on any particular square according to our new pattern, it will look like this:
+
+```bash
+bc <<< "2^($square - 1)"
+```
+
+More specifically, for the "65th" square, that would look like this:
```

Comment:
I got confused at this point why we're talking about the 65th square here. Can we either calculate the value for the 64th and final square of our chessboard, or emphasize why we're going beyond the 64 squares?

Reply:
Hopefully what I've got now is a little clearer?  

I avoided the "65th" square altogether and was still able to address the fact that the total is only the same power as the "square number" because of zero-indexing (and not magic) in a way that I'm happy with.  It really tightened up that section a lot to *show* rather than *telling* all of the little details.